### DISCUSSION TITLE
Mass Update of Price per kw/h help

### DISCUSSION THREAD

**Discussion by Oddadin** [teslamate-org/teslamate#1389] [Q&A]
Does anyone know a way to do a mass update for a geo-fenced location to amend it's price per kw/h? I put in what I thought was pence in Teslamate which then offered to update 400+ records for the geo-fence location. It turns out it was £ not pence. I can manually update them via the charges page on the Grafana console but amending the geo-fence price doesn't retro-actively change the price. Just trying to avoid manually updating all of them and my DB skills are poor at best.

**Comment by manuelkamp** [comment]
bump - I need that too since i used the wrong price per kwh as I now saw on my last-years summary bill...

**Comment by manuelkamp** [comment]
Ok, since i found out, that you can connect to the DB via this command (connect via ssh to your server where the docker container for teslamate runs):
`docker-compose exec database psql teslamate teslamate`

you can use simple SQL to update the data.
first you need to know your price per kwh. the second thing you need to know is the geofence id of the address where you charge (for my case it is id # 1 - my home address). you can find this id, by hovering in the charges tab over the location. there is a number in the URL, like this: .../geo-fences/1/edit - the number is your id.
alter now the following command with your id and your price per kwh and run it:
`UPDATE charging_processes SET cost=charge_energy_used*0.18521369 WHERE address_id='1';`
if it was successful, you recieve a message like this: UPDATE 123

enjoy your updated charging costs :)

**Reply by Oddadin** [reply]
That was spot on! I would never have figured that out so I appreciate you returning with the instructions on how to fix it.

**Reply by phioschaos** [reply]
Cool! I was looking for something like this. In my case, I did not update the cost when my utility raised the rates last year. Do you know how to add a time range to that SQL command?

**Reply by phioschaos** [reply]
Adding a date range turned out easier. Also, should use geofence_id instead of address_id. Maybe my geofence area is too small but I had home charges that had different address_id.

The command I used is:
`UPDATE charging_processes SET cost=charge_energy_used*0.18531 WHERE geofence_id='1' and start_date >= '2021-08-01 00:00:00' and start_date < '2021-11-13 00:00:00';
`
The dates are in the format year-month-day. The time is without time zone so PST would actually be 08:00:00 but this is good enough to handle updating overnight charging.

**Reply by feralbob** [reply]
> you can use simple SQL to update the data. first you need to know your price per kwh. the second thing you need to know is the geofence id of the address where you charge (for my case it is id # 1 - my home address). you can find this id, by hovering in the charges tab over the location. there is a number in the URL, like this: .../geo-fences/1/edit - the number is your id. alter now the following command with your id and your price per kwh and run it: `UPDATE charging_processes SET cost=charge_energy_used*0.18521369 WHERE address_id='1';` if it was successful, you recieve a message like this: UPDATE 123
> 
> enjoy your updated charging costs :)

I've updated the SQL a little - this looks to deal with the situation where the charge_energy_added vs charge_energy_used is different. You'll need to adjust the `geofence_id` and the `start_date` and the kWh multiplier to suit your needs.

```
WITH cost_query AS (
  SELECT id, GREATEST (charge_energy_added, charge_energy_used) AS energy
  FROM charging_processes
  WHERE geofence_id='3' and start_date >= '2022-03-01 00:00:00 America/Halifax'
)
UPDATE charging_processes
SET cost = energy*0.1532
FROM cost_query
WHERE charging_processes.id = cost_query.id
RETURNING charging_processes.id, charging_processes.cost
```

I suggest running it in a transaction just in case something goes wrong

Something like the following shows that I'm identifying the transactions that occurred after the change in kWh price, 21 of them, and successfully updating just the same 21 transactions to the new price.

```
teslamate=#  SELECT id, cost, charge_energy_added, charge_energy_used FROM charging_processes WHERE geofence_id='3' and start_date >= '2022-03-01 00:00:00 America/Halifax';
 id | cost | charge_energy_added | charge_energy_used 
----+------+---------------------+--------------------
 32 | 0.77 |                5.13 |               4.89
 33 | 3.64 |               24.39 |              22.20
 34 | 0.73 |                4.29 |               4.86
 35 | 4.36 |               29.19 |              28.32
 36 | 4.85 |               32.54 |              31.12
 38 | 2.14 |               14.33 |              14.35
 39 | 2.05 |               13.62 |              13.72
 40 | 0.00 |                0.00 |                   
 41 | 2.04 |               13.41 |              13.66
 42 | 2.19 |               14.67 |              14.56
 43 | 1.67 |                8.80 |              11.17
 44 | 0.00 |                0.00 |                   
 45 | 4.92 |               30.51 |              32.95
 46 | 3.51 |               22.95 |              23.54
 47 | 1.88 |               11.84 |              12.58
 49 | 2.01 |               12.89 |              13.47
 50 | 2.81 |               18.64 |              18.86
 51 | 0.16 |                1.04 |               1.06
 52 | 0.00 |                0.00 |                   
 53 | 3.55 |               23.67 |              23.77
 54 | 2.22 |               14.76 |              14.87
(21 rows)

teslamate=# BEGIN;
BEGIN
teslamate=*# WITH cost_query AS (
  SELECT id, GREATEST (charge_energy_added, charge_energy_used) AS energy
  FROM charging_processes
  WHERE geofence_id='3' and start_date >= '2022-03-01 00:00:00 America/Halifax'
)
UPDATE charging_processes
SET cost = energy*0.1532
FROM cost_query
WHERE charging_processes.id = cost_query.id
RETURNING charging_processes.id, charging_processes.cost;
 id | cost 
----+------
 32 | 0.79
 33 | 3.74
 34 | 0.74
 35 | 4.47
 36 | 4.99
 38 | 2.20
 39 | 2.10
 40 | 0.00
 41 | 2.09
 42 | 2.25
 43 | 1.71
 44 | 0.00
 45 | 5.05
 46 | 3.61
 47 | 1.93
 49 | 2.06
 50 | 2.89
 51 | 0.16
 52 | 0.00
 53 | 3.64
 54 | 2.28
(21 rows)

UPDATE 21

teslamate=# COMMIT;
```

Hope that helps.
